### PR TITLE
HPCC-13791 Use keys in sort partitioning

### PR DESCRIPTION
**Pull request by jakesmith**
Global sort/join prior to this change, serialized whole rows to and from the master to perform the partitioning calculation. This is particularly expensive on large clusters with large rows.
It's also very memory hungry, which has meant the global sorts of large rows have caused OOM's on the master whilst handling the partition row data.
This change serializes the sort keys only to and from master to slaves.

### CODE REVIEW

**Comment by jakesmith** [comment]
NB: This is based off and dependent on @ghalliday 's codegen changes in gh-7545

**Comment by jakesmith** [comment]
@ghalliday - please review

**Comment by richardkchapman** [comment]
Do you suppress the extra functions/serializers and the extra work (if there is any?) in the case that all fields are being compared - this is a very common case for index creation.

**Comment by ghalliday** [comment]
Yes (that is in the code generator) it only generates them if it will reduce the size.

**Comment by jakesmith** [comment]
Yep, the Thor code is conditional on the serializers being generated.

**Comment by jakesmith** [comment]
@ghalliday - repushed branch as I noticed I'd accidentally released with TBB off in commonSetup.cmake

**Comment by richardkchapman** [comment]
It SHOULD work however many ways in parallel you build - if not it suggests something is not properly marking a dependency (probably on the scm files)

**Comment by ghalliday** [comment]
@jakesmith can you rebase, now Richard has merged my PR.

**Comment by jakesmith** [comment]
@ghalliday - rebased.

File: thorlcr/activities/join/thjoin.cpp [hunk 1]
```diff
@@ -158,23 +157,56 @@ class JoinActivityMaster : public CMasterActivity
...
+                if (!rightpartition)
+                {
+                    primaryInput = container.queryInput(0);
+                    primaryCompare = helper->queryCompareLeft();
+                    primaryKeySerializer = helper->querySerializeLeft();
+                    if (container.getKind() != TAKselfjoin)
+                    {
+                        secondaryInput = container.queryInput(1);
+                        secondaryCompare = helper->queryCompareRight();
+                        secondaryKeySerializer = helper->querySerializeRight();
```
Comment (ghalliday) [inline thorlcr/activities/join/thjoin.cpp]: This will always be NULL, I now only generate a single serializer.

File: thorlcr/activities/join/thjoin.cpp [hunk 2]
```diff
@@ -158,23 +157,56 @@ class JoinActivityMaster : public CMasterActivity
...
+                if (helper->getJoinFlags()&JFslidingmatch) // JCSMORE shouldn't be necessary
+                    primaryKeySerializer = secondaryKeySerializer = NULL;
+                Owned<IRowInterfaces> primaryRowIf = createRowInterfaces(primaryInput->queryHelper()->queryOutputMeta(), queryActivityId(), queryCodeContext());
+                Owned<IRowInterfaces> secondaryRowIf;
+                if (secondaryInput)
+                    secondaryRowIf.setown(createRowInterfaces(secondaryInput->queryHelper()->queryOutputMeta(), queryActivityId(), queryCodeContext()));
+
                 bool betweenjoin = (helper->getJoinFlags()&JFslidingmatch)!=0;
-                if (!container.queryLocalOrGrouped() && container.getKind() == TAKselfjoin)
+                if (container.getKind() == TAKselfjoin)
```
Comment (ghalliday) [inline thorlcr/activities/join/thjoin.cpp]: Is this correct to lose the queryLocalOrGrouped() check?

File: thorlcr/msort/tsortm.cpp [hunk 3]
```diff
@@ -416,13 +411,15 @@ class CSortMaster: public IThorSorterMaster, public CSimpleInterface
...
         estrecsize = 100;
-        if (!partitioninfo) { // if cosort use aux
-            if (cosort) {
+        if (!partitioninfo) // if cosort use aux
+        {
+            if (cosort)
+            {
                 ActPrintLog(activity, "Cosort with no prior partition");
-                partitioninfo = new PartitionInfo(activity, auxrowif);
+                partitioninfo = new PartitionInfo(activity, keyIf);
```
Comment (ghalliday) [inline thorlcr/msort/tsortm.cpp]: trivial: two branches now do the same.

File: thorlcr/msort/tsortm.cpp [hunk 4]
```diff
@@ -441,10 +438,9 @@ class CSortMaster: public IThorSorterMaster, public CSimpleInterface
             mpp++;
         }
         if (keyserializer&&cosort&&!partitioninfo->IsOK()) {
-            keyserializer = NULL; // when joining to 0 rows can't use (LHS) serializer getMinMax will tell slave
-            ActPrintLog(activity, "Suppressing key serializer on master");
+//            keyserializer = NULL; // when joining to 0 rows can't use (LHS) serializer getMinMax will tell slave
```
Comment (ghalliday) [inline thorlcr/msort/tsortm.cpp]: Should this comment be removed/modified?

File: thorlcr/activities/join/thjoin.cpp (hunk 1 above)
Comment (jakesmith) [inline thorlcr/activities/join/thjoin.cpp]: Hm, I'm not sure all is well if LHS is already sorted, I will check/test.

File: thorlcr/activities/join/thjoin.cpp (hunk 2 above)
Comment (jakesmith) [inline thorlcr/activities/join/thjoin.cpp]: Yes, islocal will be set and it not get this far.

File: thorlcr/msort/tsortm.cpp (hunk 3 above)
Comment (jakesmith) [inline thorlcr/msort/tsortm.cpp]: Will remove.

File: thorlcr/msort/tsortm.cpp (hunk 4 above)
Comment (jakesmith) [inline thorlcr/msort/tsortm.cpp]: Probably and possibly the whole check/setting of keyserializer=NULL.
It used to cause the slave keyserialize to be muted too,
I wasn't sure what it was originally intended for, but considering the old sort code did not use the keyserializer at all, not much.. 
I should check a case where there 0 rows as a sanity check though, see if anything i haven't spotted falls out.

File: thorlcr/msort/tsortm.cpp [hunk 5]
```diff
@@ -629,32 +624,26 @@ class CSortMaster: public IThorSorterMaster, public CSimpleInterface
...
+                {
                     size32_t samplebufsize;
                     void *samplebuf=NULL;
                     slave.GetMultiNthRow(slavesamples, samplebufsize, samplebuf);
-                    CriticalBlock block(asect);
-                    sample.deserializeExpand(samplebufsize, samplebuf);
+                    MemoryBuffer mb;
+                    fastLZDecompressToBuffer(mb, samplebuf);
                     free(samplebuf);
+                    CriticalBlock block(asect);
```
Comment (ghalliday) [inline thorlcr/msort/tsortm.cpp]: Could this code be using the deserializeExpand() code that was there before, but using the keyIf instead?  Is this a change that is still needed?

File: thorlcr/msort/tsortm.cpp [hunk 6]
```diff
@@ -629,32 +624,26 @@ class CSortMaster: public IThorSorterMaster, public CSimpleInterface
...
-                if (slavesamples) {
+                if (slavesamples)
+                {
                     size32_t samplebufsize;
                     void *samplebuf=NULL;
                     slave.GetMultiNthRow(slavesamples, samplebufsize, samplebuf);
-                    CriticalBlock block(asect);
-                    sample.deserializeExpand(samplebufsize, samplebuf);
+                    MemoryBuffer mb;
+                    fastLZDecompressToBuffer(mb, samplebuf);
```
Comment (ghalliday) [inline thorlcr/msort/tsortm.cpp]: The compression should be configurable - but that can be a different pull request.

**Comment by ghalliday** [comment]
@jakesmith a few comments

File: thorlcr/msort/tsortm.cpp (hunk 5 above)
Comment (jakesmith) [inline thorlcr/msort/tsortm.cpp]: yes - it's because the rowarray has it's own serialization format (header and potentially markers for nulls) and now the slave is serializing each row through a rowToKey serializer into a buffer and compressing, i.e. they are not/can not be placed into a row array and then serialized.

File: thorlcr/msort/tsortm.cpp (hunk 6 above)
Comment (jakesmith) [inline thorlcr/msort/tsortm.cpp]: Agreed.

File: thorlcr/activities/join/thjoin.cpp (hunk 1 above)
Comment (jakesmith) [inline thorlcr/activities/join/thjoin.cpp]: I did find a problem in this area.
Fix in next commit + removal of redundant secondaryKeySerializer

**Comment by jakesmith** [comment]
@ghalliday - please see changes

File: thorlcr/activities/join/thjoin.cpp [hunk 7]
```diff
@@ -254,12 +288,12 @@ class JoinActivityMaster : public CMasterActivity
                         imaster->SortDone();
                     }
                 }
-                else { // only sort non-partition side
-                    Owned<IRowInterfaces> rowif = createRowInterfaces(container.queryInput(rightpartition?0:1)->queryHelper()->queryOutputMeta(),queryActivityId(),queryCodeContext());
-                    Owned<IRowInterfaces> auxrowif = createRowInterfaces(container.queryInput(rightpartition?1:0)->queryHelper()->queryOutputMeta(),queryActivityId(),queryCodeContext());
-                    imaster->SortSetup(rowif,rightpartition?helper->queryCompareLeft():helper->queryCompareRight(),rightpartition?helper->querySerializeLeft():helper->querySerializeRight(),true,true, NULL, auxrowif);
+                else // only sort non-partition side
+                {
+                    imaster->SortSetup(secondaryRowIf, secondaryCompare, primaryKeySerializer, true, true, NULL, primaryRowIf);
```
Comment (ghalliday) [inline thorlcr/activities/join/thjoin.cpp]: is PrimaryKeySerializer correct here?  previously I think it would have been secondary.

File: thorlcr/msort/tsortm.cpp [hunk 8]
```diff
@@ -400,14 +384,25 @@ class CSortMaster: public IThorSorterMaster, public CSimpleInterface
...
             auxrowif.set(_auxrowif);
         else
             auxrowif.set(_rowif);
         synchronized proc(slavemutex);
-        icompare = _icompare;
         keyserializer = _keyserializer;
+        if (keyserializer)
+        {
+            keyIf.setown(createRowInterfaces(keyserializer->queryRecordSize(), activity->queryContainer().queryId(), activity->queryCodeContext()));
+            keyIf.set(keyIf);
```
Comment (ghalliday) [inline thorlcr/msort/tsortm.cpp]: Harmless, but should be deleted.

**Comment by ghalliday** [comment]
@jakesmith one question, and one minor comment.

File: thorlcr/activities/join/thjoin.cpp (hunk 7 above)
Comment (jakesmith) [inline thorlcr/activities/join/thjoin.cpp]: Yes it's correct, the serializer is used to serializes partitioning keys.
Even though it is already sorted on the primary side, it is still using the primary serializer to serialize the 1st LHS row from each slave.

File: thorlcr/msort/tsortm.cpp (hunk 8 above)
Comment (jakesmith) [inline thorlcr/msort/tsortm.cpp]: Will remove

**Comment by jakesmith** [comment]
@ghalliday - removed redundant keyIf line.
